File: 2023A7PS0444G/Bitsauto/lab1_ws/src/lab1_pkg/src/talker.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from ackermann_msgs.msg import AckermannDriveStamped

class Talker(Node):
    def __init__(self):
        super().__init__('talker')

        self.declare_parameter('v', 1.0)
        self.declare_parameter('d', 0.0)
        self.publisher_ = self.create_publisher(AckermannDriveStamped, 'drive', 10)

        self.timer = self.create_timer(0.01, self.publish_drive_command)

    def publish_drive_command(self):
        v = self.get_parameter('v').get_parameter_value().double_value
        d = self.get_parameter('d').get_parameter_value().double_value

        drive_msg = AckermannDriveStamped()
        drive_msg.drive.speed = v
        drive_msg.drive.steering_angle = d

        self.publisher_.publish(drive_msg)
        self.get_logger().info(f'Publishing: speed={v}, steering_angle={d}')

def main(args=None):
    rclpy.init(args=args)
    talker = Talker()

    rclpy.spin(talker)

    talker.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: 2023A7PS0345G/2023A7PS0345G_Special_project/lab1_ws/src/lab1_pkg/src/talker.cpp
#include <chrono>
#include <functional>
#include <rclcpp/rclcpp.hpp>
#include <ackermann_msgs/msg/ackermann_drive_stamped.hpp>

using namespace std::chrono_literals;  // This line allows using the "ms" literal

class Talker : public rclcpp::Node
{
public:
    Talker()
        : Node("talker")
    {
        publisher_ = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>("drive", 10);
        timer_ = this->create_wall_timer(
            100ms, std::bind(&Talker::publish_message, this));  // Using 100ms here
    }

private:
    void publish_message()
    {
        auto message = ackermann_msgs::msg::AckermannDriveStamped();
        message.drive.speed = 0.5; // Set speed
        message.drive.steering_angle = 0.1; // Set steering angle
        publisher_->publish(message);
        RCLCPP_INFO(this->get_logger(), "Publishing: speed=%f, steering_angle=%f", message.drive.speed, message.drive.steering_angle);
    }

    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Talker>());
    rclcpp::shutdown();
    return 0;
}
